File: product-service/src/main/java/com/intexsoft/webshop/productservice/producer/impl/ProductEventProducerImpl.java
package com.intexsoft.webshop.productservice.producer.impl;

import com.intexsoft.webshop.messagecommon.event.product.impl.ProductCreatedEvent;
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductDeletedEvent;
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductOrderQuantityIncrementedEvent;
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductUpdatedEvent;
import com.intexsoft.webshop.productservice.producer.ProductEventProducer;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.core.TopicExchange;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import static com.intexsoft.webshop.productservice.util.JsonUtils.getAsString;

@Service
@RequiredArgsConstructor
@Slf4j
public class ProductEventProducerImpl implements ProductEventProducer {
    private final RabbitTemplate rabbitTemplate;
    private final TopicExchange eventExchange;
    @Value("${rmq.event.product.routing-prefix}")
    private String routingPrefix;
    @Value("${rmq.event.product.routing-keys.product_created}")
    private String productCreatedRoutingKey;

    @Value("${rmq.event.product.routing-keys.product_deleted}")
    private String productDeletedRoutingKey;

    @Value("${rmq.event.product.routing-keys.product_updated}")
    private String productUpdatedRoutingKey;
    @Value("${rmq.event.product.routing-keys.product_order_quantity_incremented}")
    private String productOrderQuantityIncrementedRoutingKey;

    @Override
    public void produceProductEventCreated(ProductCreatedEvent productCreatedEvent) {
        log.info("IN: produce {} message. The event message = {}",
                productCreatedEvent.getClass().getSimpleName(), getAsString(productCreatedEvent));
        String routing = routingPrefix + productCreatedRoutingKey;
        log.debug("Message is producing to exchange = {}, with routing = {}", eventExchange.getName(), routing);
        rabbitTemplate.convertAndSend(eventExchange.getName(), routing, productCreatedEvent);
    }

    @Override
    public void produceProductEventDeleted(ProductDeletedEvent productDeletedEvent) {
        log.info("IN: produce {} message. The event message = {}",
                productDeletedEvent.getClass().getSimpleName(), getAsString(productDeletedEvent));
        String routing = routingPrefix + productDeletedRoutingKey;
        log.debug("Message is producing to exchange = {}, with routing = {}", eventExchange.getName(), routing);
        rabbitTemplate.convertAndSend(eventExchange.getName(), routing, productDeletedEvent);
    }

    @Override
    public void produceProductEventUpdated(ProductUpdatedEvent productUpdatedEvent) {
        log.info("IN: produce {} message. The event message = {}",
                productUpdatedEvent.getClass().getSimpleName(), getAsString(productUpdatedEvent));
        String routing = routingPrefix + productUpdatedRoutingKey;
        log.debug("Message is producing to exchange = {}, with routing = {}", eventExchange.getName(), routing);
        rabbitTemplate.convertAndSend(eventExchange.getName(), routing, productUpdatedEvent);
    }

    @Override
    public void produceProductOrderQuantityIncremented(ProductOrderQuantityIncrementedEvent productOrderQuantityIncrementedEvent) {
        log.info("IN: produce {} message. The event message = {}",
                productOrderQuantityIncrementedEvent.getClass().getSimpleName(), getAsString(productOrderQuantityIncrementedEvent));
        String routing = routingPrefix + productOrderQuantityIncrementedRoutingKey;
        log.debug("Message is producing to exchange = {}, with routing = {}", eventExchange.getName(), routing);
        rabbitTemplate.convertAndSend(eventExchange.getName(), routing, productOrderQuantityIncrementedEvent);
    }
}

File: product-service-kt/src/main/kotlin/com/intexsoft/webshop/productservicekt/producer/impl/ProductEventProducerImpl.kt
package com.intexsoft.webshop.productservicekt.producer.impl

import com.intexsoft.webshop.messagecommon.event.BaseEvent
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductCreatedEvent
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductDeletedEvent
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductOrderQuantityIncrementedEvent
import com.intexsoft.webshop.messagecommon.event.product.impl.ProductUpdatedEvent
import com.intexsoft.webshop.productservicekt.log
import com.intexsoft.webshop.productservicekt.producer.ProductEventProducer
import org.springframework.amqp.core.TopicExchange
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
class ProductEventProducerImpl(
    private val rabbitTemplate: RabbitTemplate,
    private val eventExchange: TopicExchange
) : ProductEventProducer {


    @Value("\${rmq.event.product.routing-prefix}")
    private lateinit var routingPrefix: String

    @Value("\${rmq.event.product.routing-keys.product_created}")
    private lateinit var productCreatedRoutingKey: String

    @Value("\${rmq.event.product.routing-keys.product_deleted}")
    private lateinit var productDeletedRoutingKey: String

    @Value("\${rmq.event.product.routing-keys.product_updated}")
    private lateinit var productUpdatedRoutingKey: String

    @Value("\${rmq.event.product.routing-keys.product_order_quantity_incremented}")
    private lateinit var productOrderQuantityIncrementedRoutingKey: String
    override fun produceProductEventCreated(productCreatedEvent: ProductCreatedEvent) {
        val routing = routingPrefix + productCreatedRoutingKey
        produceEventWithRouting(productCreatedEvent, routing)
    }

    override fun produceProductEventDeleted(productDeletedEvent: ProductDeletedEvent) {
        val routing = routingPrefix + productDeletedRoutingKey
        produceEventWithRouting(productDeletedEvent, routing)
    }

    override fun produceProductEventUpdated(productUpdatedEvent: ProductUpdatedEvent) {
        val routing = routingPrefix + productUpdatedRoutingKey
        produceEventWithRouting(productUpdatedEvent, routing)
    }

    override fun produceProductOrderQuantityIncremented(productOrderQuantityIncrementedEvent: ProductOrderQuantityIncrementedEvent) {
        val routing = routingPrefix + productOrderQuantityIncrementedRoutingKey
        produceEventWithRouting(productOrderQuantityIncrementedEvent, routing)
    }

    private fun produceEventWithRouting(baseEvent: BaseEvent, routing: String) {
        log.info("IN: produce ${baseEvent::class.simpleName} message. The event message = $baseEvent")
        log.debug("Message is producing to exchange = ${eventExchange.name}, with routing = $routing")
        rabbitTemplate.convertAndSend(eventExchange.name, routing, baseEvent)
    }
}